core/shared/src/main/scala/fs2/concurrent/Channel.scala

```diff
@@ -209,7 +209,8 @@ object Channel {
                     (
                       State(values, size, waiting.some, producers, closed),
                       F.pure(
-                        (Pull.eval(waiting.get) >> consumeLoop).unlessA(closed)
```

Comment:
if you have a `Defer[Pull]` which I guess you should be able to implement, then you could also write this as:

`Defer[Pull].defer(Pull.eval(waiting.get) >> consumeLoop).unlessA(closed)`

although explicitly putting the branch here to get the result you want is probably clearer (although I would probably comment it so no one refactors it later and regresses the bug).

Reply:
I hope it gets fixed upstream in cats as well